javascript/ql/src/experimental/Security/CWE-117/LogInjection.qll

```diff
@@ -0,0 +1,105 @@
+/**
+ * Provides a taint-tracking configuration for reasoning about untrusted user input used in log entries.
+ */
+
+import javascript
+
+module LogInjection {
+  /**
+   * A data flow source for user input used in log entries.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for user input used in log entries.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer for malicious user input used in log entries.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A taint-tracking configuration for untrusted user input used in log entries.
+   */
+  class LogInjectionConfiguration extends TaintTracking::Configuration {
+    LogInjectionConfiguration() { this = "LogInjection" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isSanitizer(DataFlow::Node node) { node instanceof Sanitizer }
+  }
+
+  /**
+   * A source of remote user controlled input.
+   */
+  class RemoteSource extends Source {
+    RemoteSource() { this instanceof RemoteFlowSource }
+  }
+
+  /**
+   * An source node representing a logging mechanism.
+   */
+  class ConsoleSource extends DataFlow::SourceNode {
+    ConsoleSource() {
+      exists(DataFlow::SourceNode node |
+        node = this and this = DataFlow::moduleImport("console")
+        or
+        this = DataFlow::globalVarRef("console")
+      )
+    }
+  }
+
+  /**
+   * A call to a logging mechanism. For example, the call could be in the following forms:
+   * `console.log('hello')` or
+   *
+   * `let logger = console.log;`
+   * `logger('hello')` or
+   *
+   * `let logger = {info: console.log};`
+   * `logger.info('hello')`
+   */
+  class LoggingCall extends DataFlow::CallNode {
+    LoggingCall() {
+      this = any(ConsoleSource console).getAMemberCall(getAStandardLoggerMethodName())
+      or
+      exists(DataFlow::SourceNode node, string propName |
+        any(ConsoleSource console).getAPropertyRead() = node.getAPropertySource(propName) and
+        this = node.getAPropertyRead(propName).getACall()
+      )
+      or
+      this = any(LoggerCall call)
```

Comment:
This is mostly a comment.

This last conjunct should cover the two cases above if https://github.com/github/codeql/blob/7a5aae7432be58fa27924a814ef3c44dd7e99512/javascript/ql/src/semmle/javascript/frameworks/Logging.qll#L57 was ` console().getAMemberCall(name)`.

(I think `console.log` used to require a binding with `Function.prototype.bind` in order for it to work when invoked as a non-method call, but that appears to have changed now).

Reply:
Thank you for your review and feedback.

Yes, if it would be `console().getAMemberCall(name)` I could avoid the first case
`this = any(ConsoleSource console).getAMemberCall(getAStandardLoggerMethodName())`

but not the second one.

The second check
```
      exists(DataFlow::SourceNode node, string propName |
        any(ConsoleSource console).getAPropertyRead() = node.getAPropertySource(propName) and
        this = node.getAPropertyRead(propName).getACall()
      )
```
detects cases where the `console.log` (for example) is assigned to a property of another object.
For example, consider the following case:
```
const my_logger = {
    log: console.log
}
my_logger.log("hi")
```
Though the `Logging.dll` will be updated to `console().getAMemberCall(name)`, I will still miss the above case (please, do not hesitate to correct me if I am wrong).

Reply:
I have opened #3767 to address the first case, so lets drop that case in this PR.

For the second case, we have a bit better machinery for tracking special values of interest through properties and even calls ("type tracking"), but lets just keep the second case as is.
Let me know if you want to try out the "type tracking" feature later.

Reply:
Sorry for the late reply. I am really happy that I inspire the other PR.
Thank you again for your feedback.

Since now it is merged, I remove the first case, and leave as is the second one.

Let me know if there is something I can do.

About the "type tracking" feature  is something that I will investigate.